### DISCUSSION TITLE
Non-Max Supression Issue

### DISCUSSION THREAD

**Discussion by MohitDAngrish** [facebookresearch/detectron2#3484] [Q&A]
After training the detectron faster_rcnn_R_50_C4_3x for a certain usecase with the given below config :

```
cfg = get_cfg()
cfg.merge_from_file(model_zoo.get_config_file("COCO-Detection/faster_rcnn_R_50_C4_3x.yaml"))
cfg.DATASETS.TRAIN = ("my_dataset_train_new",)
cfg.DATASETS.TEST = ("my_dataset_test_new",)
cfg.DATALOADER.NUM_WORKERS = 4
cfg.MODEL.WEIGHTS = model_zoo.get_checkpoint_url("COCO-Detection/faster_rcnn_R_50_C4_3x.yaml")  # Let training initialize from model zoo
cfg.SOLVER.IMS_PER_BATCH = 4
cfg.SOLVER.MAX_ITER = 60000 #adjust up if val mAP is still rising, adjust down if overfit
cfg.SOLVER.WARMUP_ITERS = 1000
cfg.SOLVER.LR_SCHEDULER_NAME = 'WarmupCosineLR'
cfg.SOLVER.STEPS = []
cfg.OUTPUT_DIR = '/content/drive/MyDrive/ObjectDetectionDetectron/Models/faster_rcnn_R_50_C4_3xBestModelVal'
cfg.SOLVER.GAMMA = 0.05
cfg.MODEL.ROI_HEADS.BATCH_SIZE_PER_IMAGE = 64
cfg.MODEL.ROI_HEADS.NUM_CLASSES = 8
cfg.TEST.EVAL_PERIOD = 20
cfg.MODEL.ROI_HEADS.NMS_THRESH_TEST= 0.6 
```

**Train loss** is reducing upto **0.3** but the **val loss** is still at **0.9**. Model seem to be overfitting. But that's not the case when i try doing predictions the output seemed reasonable but there was an issue of overlapping boxes as you can see below.

![test](https://user-images.githubusercontent.com/26500540/133495644-ac134f95-1eff-4142-b5a6-e1ad10513650.png)

So here are my two questions:
1. If **NMS** and **IOU** is not working why the train loss is so low.
2. If **NMS** and **IOU** is working can somebody please point me towards the code where the non max suppression and iou is handled while training and evaluation of a model is done.
